Improve network transport

dnkolegov opened the pull request.

matejpavlovic reviewed (approved):
Good job! Most comments are just cosmetic.
Thanks! :-)

dnkolegov commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -105,84 +113,122 @@ func (t *Transport) Stop() {
 	} else {
 		t.logger.Log(logging.LevelDebug, fmt.Sprintf("node %v libp2p host closed", t.ownID))
```
Fixed

dnkolegov commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -290,23 +344,84 @@ func (t *Transport) mirHandler(s network.Stream) {
+	return found && info.IsOpening
+}
+
+func (t *Transport) addConnectionInProgress(nodeID types.NodeID) {
```
Fixed

dnkolegov commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -105,84 +113,122 @@ func (t *Transport) Stop() {
 	s, err := t.openStream(ctx, info.ID)
 	if err != nil {
-		return nil, fmt.Errorf("failed to open new stream to node %v: %w", addr, err)
+		t.removeConnectionInProgress(node)
```
Good catch! removed!

dnkolegov commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -219,38 +265,46 @@ func (t *Transport) openStream(ctx context.Context, p peer.ID) (network.Stream,
-	if err != nil {
-		return fmt.Errorf("failed to marshal message: %w", err)
+func (t *Transport) Send(ctx context.Context, dest types.NodeID, payload *messagepb.Message) error {
+	found := t.nodeExists(dest)
+	if !found {
+		return fmt.Errorf("failed to get address for node %s", dest)
 	}
```
The difference is that if the target node is not found it is not possible to add this node within libp2p. And because of that, it is not possible to open a stream. Only the caller can fix it somehow (e.g., via Connect function).

Does it make sense?

>  handled by t.connectToNode.
That's true. But `connectToNode` doesn't return an error. And because of that and async nature of that function, we will run
```
go t.connectToNode(ctx, dest, t.connWg)
		return fmt.Errorf("no open stream for node %s, reopening", dest)
```
Then we get an error "that node doesn't exist" in logs and get "no open stream for node" error in the caller function.

As a result, I think we should distinguish between cases when the node doesn't exist at all and the case when the node exists but the stream doesn't

matejpavlovic replied:
Yes I agree that this potentially might be a useful thing to distinguish, but unless it happens atomically (checking both things while holding the lock), one can always construct a scenario where the first check is futile.
E.g., the check succeeds, releases the lock, and then someone removes the node by calling `CloseOldConnections`.

If we do want to distinguish it properly, I think the check still needs to happen atomically (for both the presence of the nodeInfo and the connection within). For example by extending the `t.getStream` function so we can make this distinction based on what it returns.

In fact, in any case, it still seems possible that the stream is closed *concurrently* by `CloseOldConnections` when `Send` is writing to it. It would be good to check with the doc that this doesn't panic and behaves in a predictable way.

dnkolegov replied:
Agree. Fixed

dnkolegov commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -290,23 +344,84 @@ func (t *Transport) mirHandler(s network.Stream) {
-func (t *Transport) streamExists(nodeID types.NodeID) bool {
-	t.outboundStreamsMx.Lock()
-	defer t.outboundStreamsMx.Unlock()
+func (t *Transport) nodeExists(nodeID types.NodeID) bool {
```
Fixed

matejpavlovic commented:
Nice, thanks for the fixes! Feel free to merge when you make the tests pass :-)

jsoares commented:
Continuation of Slack thread wrt to connection retries: we seem to have an answer wrt to libp2p behaviour (https://github.com/libp2p/go-libp2p/blob/7828f3e0797e0a7b7033fa5e8be9b94f57a4c173/p2p/net/swarm/swarm.go#L358, per Alfonso). Is the path clear on our end, or do we still need input from the libp2p team?

cc @dnkolegov @adlrocha @vukolic @sergefdrv

dnkolegov commented:
Yes, probably, we have an answer to the question.

>  Is the path clear on our end, or do we still need input from the libp2p team?

@jsoares I would be happy if someone from the libp2p team reviews this PR! Again, having in mind that Mir is going to the production.

sergefdrv commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,492 @@
+	w := bufio.NewWriter(s)
+	_, err = w.Write(buf.Bytes())
+	if err == nil {
+		w.Flush()
```
IIUC, `Flush` may write the buffered data only partially and return `io.ErrShortWrite` ([source](https://cs.opensource.google/go/go/+/refs/tags/go1.19.1:src/bufio/bufio.go;drc=99f1bf54eb502e123e60c605212a616146fbe86a;l=631)). Have you checked if libp2p stream writer may write less bytes than requested?

dnkolegov replied:
Yes, good catch. That is a debt from the initial implementation. Fixed

sergefdrv replied:
I wonder if libp2p itself can write the data partially and return `io.ErrShortWrite` without the stream being broken

matejpavlovic replied:
Can a short write happen even without the connection breaking? If that's the case, we might need to forcefully break the connection and create a new one (or somehow else tell the other end that whatever comes next is a new message, not the continuation of the partially sent one)...

dnkolegov replied:
Created an issue to address this - https://github.com/filecoin-project/mir/issues/255

adlrocha replied:
I don't know IIU the issue correctly, but I honestly don't think this is possible. Libp2p expects the buffer to be explicitly flushed so it can notify the length of the stream being sent to other side (all stream communications are prefixed by a `VarInt` notifying the receiver the amount of bytes to read from the buffer, and buffers are reused on top of the existing connections).

sergefdrv commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,492 @@
+	w := bufio.NewWriter(s)
+	_, err = w.Write(buf.Bytes())
+	if err == nil {
+		w.Flush()
```
This call will block if the underlying writer is busy. I'm afraid that any destination node could cause the whole module to block here.

matejpavlovic replied:
Yes, that is a good point too. We should make sure that a single slow connection doesn't block all the other ones. A proper solution to this might be actually non-trivial, probably involving separate per-connection output buffers, each operated by its own goroutine, and dropping outgoing messages when they become full... But maybe we can address this in a separate PR.

dnkolegov replied:
Created an issue to address this - https://github.com/filecoin-project/mir/issues/255

sergefdrv commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,494 @@
+	w := bufio.NewWriter(s)
+	_, err = w.Write(buf.Bytes())
+	if err == nil {
+		if err = w.Flush(); err != nil {
```
It may still write data partially to the libp2p stream. What happens on the other end in that case?

dnkolegov replied:
Created an issue to address this - https://github.com/filecoin-project/mir/issues/255

sergefdrv commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,494 @@
+			return
+		}
+
+		t.incomingMessages <- events.ListOf(
+			events.MessageReceived(types.ModuleID(payload.DestModule), types.NodeID(msg.Sender), &payload),
+		)
```
What if the destination module does not exist?

dnkolegov replied:
Do you mean what if the destination module doesn't read from the channel?
Or what will Mir do if it gets a message to a module that doesn't exist?

sergefdrv replied:
The latter

dnkolegov replied:
@matejpavlovic Do you see any problems with that? I think that the network transport should know about that and assume that Mir properly handles that.

matejpavlovic replied:
Yes, this is not the Net module's responsibility. The node must handle the situation. This is part of input sanitization.
While complex input is a bigger issue, I've just addressed this particular part in #253 . So I think this comment can be marked as resolved.

sergefdrv replied:
I think one can make an argument that it's the module's fault emitting events which target a wrong module.

adlrocha commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,500 @@
+	}
+	t.nodesLock.Unlock()
+
+	t.connWg.Add(1)
```
Why are you using this `WaitGroup` here, instead of including all this waiting logic directly in `connect` (It feels harder to reason about if the waiting logic is scattered through different functions)? Can the high-level function `Connect` be called by several routines?

dnkolegov replied:
Agree, I moved this inside of `connect` function

adlrocha commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,500 @@
+	t.nodesLock.Unlock()
+
+	t.connWg.Add(1)
+	go t.connect(ctx, nodes)
```
Do we really need this in a routine? We are already spawning new routines for each node when we call `connect`.

dnkolegov replied:
We need if we want to not wait when all connecting goroutines stop, otherwise `wg.Wait` blocks us.

adlrocha commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,500 @@
+	}
+	// If we cannot write to the stream, or we don't have a connection to that peer then try to
+	// connect to the peer.
+	if err != nil || len(t.host.Network().ConnsToPeer(s.Conn().RemotePeer())) == 0 {
```
We should have a connection here by now, right? You already called `connectToNode` above so this is a sanity-check, but do we really want to try to connect again with the node if the stream failed? Maybe is just better to restart the connection (`Close() + Open()).

Also we should lock the stream while we write on it to prevent parallel writes. Instead of just locking in `getNodeStream` we should have a lock also while we write on it.

dnkolegov replied:
> You already called connectToNode above

That was called in case we don't have a stream to write in. 
But what if we have already had a stream, we tried to write in it, and get an error?
But I agree we should close the connection!

adlrocha commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,500 @@
+		node.Stream == nil {
+		return nil, &nilStreamError{SrcNode: t.ownID, DestNode: nodeID}
+	}
+	return node.Stream, nil
```
As mentioned above, Streams should be locked while being written to avoid parallel accesses.

dnkolegov replied:
Yes, good catch! I combined two functions and forgot to change locking

adlrocha commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,500 @@
+					// Send message to another node.
+					if err := t.Send(ctx, types.NodeID(destID), e.SendMessage.Msg); err != nil { // nolint
+						// TODO: Handle sending errors (and remove "nolint" comment above).
+						//       Also, this violates the non-blocking operation of ApplyEvents method. Fix it.
```
I don't get this. Isn't this a "fire-and-forget"? We don't wait for a response, do we?

dnkolegov replied:
Fixed. It was legacy code

adlrocha commented on pkg/net/libp2p/libp2p_transport_test.go:
```diff
@@ -0,0 +1,309 @@
+	b.Connect(ctx, initialNodes)
+	c.Connect(ctx, initialNodes)
+
+	m.testEventuallyConnected(nodeA, nodeB)
```
Should we also test if the retries work and have some parallel tests (with calls from different routines)? I haven´t looked in depth into the code, so maybe they are already there.

matejpavlovic replied:
(This applies to some of the previous comments too.)

In the current implementation, there is no way a connection could be written to concurrently by multiple threads. The whole module is handled by a single thread, sending messages one by one. This is also why getting stuck on writing to one broken (or slow) connection is such a big problem. However, at least we don't have to be concerned about locking the streams when writing into them (unless we need to synchronize the writes with potential reconnections).

dnkolegov replied:
We have a basic check that retry works - https://github.com/filecoin-project/mir/blob/be26a6d97e5d97d0c2a87ba4056f5d16cfc0a58f/pkg/net/libp2p/libp2p_transport_test.go#L235

adlrocha reviewed (changes requested):
Overall, LGTM. Just a request for change and a suggestion:
- We need to lock writes to the streams so several routines don´t try to write at the same time to the same stream.
- It is a bit hard to reason about the correctness of the concurrency in the connection manager. I wouldn't recommend having an active `WaitGroup` scattered through different functions if we can avoid it. It makes the code harder to read and it may lead to deadlocks (but maybe is just personal taste, so take this with a grain of salt).

adlrocha commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,463 @@
+}
+
+func (t *Transport) getNodeStreamAndInfo(nodeID types.NodeID) (network.Stream, *peer.AddrInfo, error) {
+	t.nodesLock.Lock()
```
This is not enough, you need to keep the lock while you write the stream. I would:
- Make this function `getNodeStreamAndInfoWithoutLock`.
- Use this function when getting the stream to write by locking before getting and releasing after flush.
- Create a `getNodeStreamAndInfoWithLock` and keep the `RLock` you had before.

Reference: https://github.com/filecoin-project/mir/pull/230/commits/1667e25c53c50b79e0ab4c9d2a9bb9ba5b4284a0

dnkolegov replied:
Fixed

sergefdrv commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,484 @@
+		return err
+	}
+
+	w := bufio.NewWriter(stream)
+	if _, err = w.Write(payload); err != nil {
+		return err
+	}
+	return w.Flush()
```
What is the advantage of using `bufio.Writer` here?

dnkolegov replied:
That is a common [approach](https://github.com/libp2p/go-libp2p/tree/master/examples) for libp2p applications.

dnkolegov replied:
What would you recommend to use instead of that?

sergefdrv replied:
Why not writing to the libp2p stream directly?

dnkolegov replied:
fixed

sergefdrv commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,484 @@
+}
+
+func (t *Transport) connect(ctx context.Context, nodes map[types.NodeID]types.NodeAddress) {
+	t.connWg.Add(1)
+	defer t.connWg.Done()
```
This pair of lines doesn't seem to make sense: the subsequent go statement will return right away causing the defer statement to execute immediately.

dnkolegov replied:
Fixed. `defer t.connWg.Done()` must be in the goroutine

sergefdrv commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,484 @@
+func (t *Transport) connectToNode(ctx context.Context, node types.NodeID, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	if t.isConnectionToNodeInProgress(node) {
+		t.logger.Log(logging.LevelDebug, "connecting to node is in progress", "src", t.ownID, "dst", node)
+		return
+	}
+
+	t.setConnectionInProgress(node)
+	defer t.clearConnectionInProgress(node)
```
Are those calls supposed to make some synchronization scheme up or they are just there for reducing the likelihood of duplicate attempt to establish the same connection? I think there is a possible race condition since `t.nodesLock` gets unlocked in-between.

dnkolegov replied:
Fixed

sergefdrv commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,484 @@
+			return
+		}
+		t.incomingMessages <- events.ListOf(
+			events.MessageReceived(types.ModuleID(msg.DestModule), sender, msg),
```
I don't think we can trust `sender` here, since we receive it from the other, potentially faulty, node and doesn't seem to verify in any way.

dnkolegov replied:
I can do the following:
```
                msg, sender, err := t.decode(s)
		if err != nil {
			t.logger.Log(logging.LevelError, "failed to read mir request", "src", t.ownID, "sender", sender, "err", err)
			return
		}
		t.nodesLock.RLock()
		v, ok := t.nodes[sender]
		t.nodesLock.RUnlock()
		if !ok {
			return
		}
		msgSenderID := v.AddrInfo.ID.String()
		senderPeerID := s.Conn().RemotePeer().String()
		if msgSenderID != senderPeerID {
			t.logger.Log(logging.LevelWarn, 
				"failed to validate sender",
				"src", t.ownID, "sender", sender, "msgSender", msgSenderID, "peerID", senderPeerID)
			return
		}
		t.incomingMessages <- events.ListOf(
			events.MessageReceived(types.ModuleID(msg.DestModule), sender, msg),
		)
```

dnkolegov replied:
Fixed

sergefdrv replied:
Now we have some additional overhead for every received message. We may want to improve this in future.

sergefdrv commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,484 @@
+var ErrNilStream = errors.New("stream has not been opened")
+
+type nodeInfo struct {
+	ID        types.NodeID
+	Addr      types.NodeAddress
```
`ID` and `Addr` don't seem to be used.

sergefdrv commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,484 @@
+				continue
+			}
+
+			if t.streamExists(nodeID) {
+				t.logger.Log(logging.LevelInfo, "stream to node already exists", "src", t.ownID, "dst", nodeID)
+				wg.Done()
+				continue
+			}
+
+			go t.connectToNode(ctx, nodeID, wg)
```
`connectToNode` doesn't seem to re-check if the stream is already exist with `nodesLock` held. I think there is a possible race condition.

dnkolegov replied:
Do you want to check that again?

sergefdrv replied:
Not necessarily, we just need to make sure no race condition happens here.

dnkolegov commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,484 @@
+
+		for nodeID, dest := range t.nodes {
+			// Close an old connection to a node if we don't need to connect to it further.
+			if _, foundInNewNodes := newNodes[nodeID]; !foundInNewNodes {
```
I don't see a reason to do it in the current network addressing in Mir since all nodes are identified by NodeIDs.

dnkolegov replied:
We don't support this case now. But if we did then we would need to answer the following question: what if we have a connection to nodeID and receive the same nodeID with a different address? what should we do? Do we need to establish a new connection and remove the old one, to support both? Until we don't have a clear requirement and use case for that let's use the simplest approach.

sergefdrv replied:
Right, we don't yet support updating the validator network address in the subnet actor. I think we should reverse the direction of communication between nodes such that the node receives messages only from those peers which it initiated connections to. Similarly, incoming connections would be served by primarily supplying messages there. Nodes could reject incoming connections from nodes with unexpected identities.

sergefdrv replied:
> what if we have a connection to nodeID and receive the same nodeID with a different address? what should we do? Do we need to establish a new connection and remove the old one, to support both?

I think we should interpret this as a simple network endpoint update and replace the old connection with a new one.

dnkolegov replied:
Then I would suggest addressing it later.
https://github.com/filecoin-project/mir/issues/259

dnkolegov commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,484 @@
+	t.connWg.Wait()
+}
+
+func (t *Transport) CloseOldConnections(newNodes map[types.NodeID]types.NodeAddress) {
```
@matejpavlovic what do you think? do we need this?

dnkolegov replied:
Will be addressed [here](https://github.com/filecoin-project/mir/issues/255). The most important question is to define the requirement!

adlrocha commented:
Quick question before we land this PR. While integrating Mir I realized that we are using a single node address to contact peers, but peers (and specifically libp2p hosts) may expose more than one endpoint. Do we want to support this? (I think we should but I don't know how hard it is to include, if it is hard lets track on an issue and defer to the future). 

@matejpavlovic @dnkolegov thoughts? 🙏

dnkolegov commented:
> Quick question before we land this PR. While integrating Mir I realized that we are using a single node address to contact peers, but peers (and specifically libp2p hosts) may expose more than one endpoint. Do we want to support this? (I think we should but I don't know how hard it is to include, if it is hard lets track on an issue and defer to the future).
> 
> @matejpavlovic @dnkolegov thoughts? 🙏

I would prefer to defer to the future. The current implementation assumes basic Mir's membership model - nodeID:nodeAddress

dnkolegov commented:
@sergefdrv @matejpavlovic @adlrocha 

I simplified the implementation.
Now Connect, CloseOldConnections, and reconnect can not be run simultaneously. But Send can work in parallel with all these functions. Let me know what you think.

Could you also resolve the conversations if you believe I have implemented your suggestions or if it is not actual anymore?

sergefdrv commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,452 @@
+type connInfo struct {
+	AddrInfo  *peer.AddrInfo
+	Stream    network.Stream
+	IsOpening bool
```
```suggestion
```

dnkolegov replied:
Fixed

sergefdrv commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,452 @@
+			t.connsLock.Lock()
+			defer t.connsLock.Unlock()
+
+			t.connWg.Add(1)
+			t.connectToNodeWithoutLock(ctx, node, t.connWg)
```
Now that you don't start a new goroutine here, this usage of `t.connWg` looks superfluous.

dnkolegov replied:
Agree, but we use the same function and to do that we need a wg.
And we can't just call `go t.connectToNodeWithoutLock(ctx, node, t.connWg)` due to Lock/Unlock

sergefdrv commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,452 @@
+
+	t.host.RemoveStreamHandler(ProtocolID)
+
+	t.connsLock.Lock()
+	for nodeID, addr := range t.conns {
+		if err := t.host.Network().ClosePeer(addr.AddrInfo.ID); err != nil {
+			t.logger.Log(logging.LevelError, "could not close old connection to node", "src", t.ownID, "dst", nodeID, "err", err)
+			continue
+		}
+	}
+	t.connsLock.Unlock()
```
Does this prevent reconnection right afterwards?

dnkolegov replied:
No, it doesn't. If you call `Stop` and then call `Connect` then new outbound connections can be established but new inbound connections to Mir protocol can't.
We can't close the `host` since we don't own it.

sergefdrv replied:
I mean `Stop` and `reconnect` methods running concurrently.

sergefdrv commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,452 @@
+	if !found {
+		t.logger.Log(logging.LevelError, "failed to get node address", "src", t.ownID, "dst", node)
+	} else {
+		if err := t.host.Network().ClosePeer(info.ID); err != nil {
+			t.logger.Log(logging.LevelError, "could not close the connection to node", "src", t.ownID, "dst", node, "err", err)
+		}
```
Perhaps we should also protect this with `t.connsLock`

dnkolegov replied:
t.connsLock is used to protect t.conns. `t.host.Network().ClosePeer` doesn't change or read `t.conns`elements

dnkolegov replied:
No. `t.conns` is just information about address and streams

dnkolegov replied:
Added removing address from t.conns when calling `Stop`

sergefdrv replied:
Well, we get `t.conns[node].Stream` from `t.host`. I was wondering if calling `t.host.Network().ClosePeer` invalidates the corresponding stream as a side-effect.

sergefdrv replied:
Ok, I see that `sendPayload` and `reconnect` cannot run concurrently for the same node.

sergefdrv commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,452 @@
+		sctx, cancel := context.WithTimeout(ctx, maxConnectingTimeout)
+
+		s, err = t.host.NewStream(sctx, dest, ProtocolID)
+		cancel()
```
Why do you call `cancel` here?

dnkolegov replied:
Because that is context.WithTimeout's cancel. 
```
Canceling this context releases resources associated with it, so code should call cancel as soon as the operations running in this
```

sergefdrv commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,452 @@
+	}
+}
+
+func (t *Transport) getNodeStreamAndInfoWithoutLock(nodeID types.NodeID) (network.Stream, error) {
```
```suggestion
func (t *Transport) getNodeStreamWithoutLock(nodeID types.NodeID) (network.Stream, error) {
```

dnkolegov replied:
Fixed

sergefdrv commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,452 @@
+	t.logger.Log(logging.LevelDebug, "stopping libp2p transport", "ownID", t.ownID)
+	defer t.logger.Log(logging.LevelDebug, "stopping libp2p transport finished.", "ownID", t.ownID)
+
+	t.host.RemoveStreamHandler(ProtocolID)
```
Does this also close incoming connections?

dnkolegov replied:
It doesn't close connections directly
https://github.com/multiformats/go-multistream/blob/225d479012e89db2156d325a60b9374713f21769/multistream.go#L148

sergefdrv replied:
Then I think we'd better close all the streams in `mirHandler`.

dnkolegov replied:
`mirHandler` is the same as a handler in`http`. I don't see how to implement that.

sergefdrv replied:
It receives `network.Stream` which has `Close` method. I think we should arrange all loops running in `mirHandler` to call `s.Close` once we call `Stop`.

sergefdrv replied:
The difference from an HTTP handler is that `mirHandler` has a long-running loop inside. If we stop the transport module, it should also drop the incoming connections, otherwise the corresponding resources cannot be garbage-collected.

dnkolegov replied:
That shows that it doesn't close connections.

dnkolegov replied:
Probably, I did not understand you right.

sergefdrv replied:
I suspect that neither `RemoveStreamHandler` nor `Network().ClosePeer` would close an already established incoming stream connection.

sergefdrv replied:
Stream and/or connection :)

sergefdrv replied:
> ClosePeer closes the connection to a given peer

Here I'm not sure it also closes incoming connections.

adlrocha replied:
FWIW (and I just skimmed through the discussion), I usually close all streams before closing the host as a sanity check with something like:
```go 
// Unregister handler
...
// Close streams
for _, s := range streams{
   s.Close()
}
// Close host
h.Close()
```

dnkolegov replied:
If I understood Sergay correctly he suggests closing all streams in the handler. 
Currently, the handler closes the stream on defer.

sergefdrv replied:
Yes, on `Stop`, I suggest closing all the streams received in `mirHandler` and waiting until all handlers exit.

adlrocha commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,452 @@
+				continue
+			}
+
+			go t.connectToNodeWithoutLock(ctx, nodeID, wg)
```
I think we are missing a `wg.Done()` below this line.

adlrocha commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,452 @@
+}
+
+func (t *Transport) connectToNodeWithoutLock(ctx context.Context, nodeID types.NodeID, wg *sync.WaitGroup) {
+	defer wg.Done()
```
```suggestion
	defer wg.Done()
```

This is not needed anymore right? You handle the waiting groups in the functions where they are called, mainly `reconnect` and `connect`

adlrocha reviewed (approved):
Ready to merge!

matejpavlovic commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,459 @@
+	}()
+}
+
+func (t *Transport) reconnect(ctx context.Context, node types.NodeID) {
```
Related to previous comment: Is there a fundamental difference between `connect` and `reconnect`, except the former does it for multiple nodes in a batch?

dnkolegov replied:
Fixed

matejpavlovic commented on pkg/util/libp2p/libp2p.go:
```diff
@@ -31,6 +32,24 @@ func NewDummyHost(numericID int, sourceAddr t.NodeAddress) host.Host {
+		var l *net.TCPListener
+		if l, err = net.ListenTCP("tcp", a); err == nil {
+			defer l.Close() // nolint
+			addr := l.Addr().(*net.TCPAddr).Port
```
Cosmetic: Calling a port `addr` is a bit deceiving.

dnkolegov replied:
fixed

matejpavlovic reviewed (commented):
I added some more comments, but some of them I suggest to address in the next PR. Will att them to #255 .

dnkolegov commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,459 @@
+			t.logger.Log(logging.LevelError, "failed to read mir message", "src", t.ownID, "err", err)
+			return
+		}
+		t.connsLock.RLock()
+		senderNode, ok := t.conns[sender]
+		t.connsLock.RUnlock()
+		if !ok {
+			return
+		}
```
I don't think so. It even doesn't return if you close the host.

dnkolegov commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,459 @@
+	}
+	t.connsLock.Unlock()
+
+	t.connect(ctx, nodes)
```
Fixed.

dnkolegov commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,459 @@
+	}() // nolint
+
+	for {
+		msg, sender, err := t.readAndDecode(s)
```
What "preemptable call" is? The call using gosched?

dnkolegov replied:
Added stopChan. Did you mean something like that? if you did not let's discuss

matejpavlovic replied:
Yes, by "preemptable" I meant that we must be able to interrupt the waiting for the data. The `stopChan` is a way of achieving that (since when `stopChan` is closed, the stream is closed, which makes the unmarshalling return, which in turn makes `readAndDecode` return).

dnkolegov commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,459 @@
+
+func (t *Transport) readAndDecode(s io.Reader) (*messagepb.Message, types.NodeID, error) {
+	var tm TransportMessage
+	err := tm.UnmarshalCBOR(s)
```
Addressed via `stopChan`

dnkolegov commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,459 @@
+			return
+		}
+
+		t.incomingMessages <- events.ListOf(
```
Fixed

dnkolegov commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,459 @@
+				continue
+			}
+
+			go t.connectToNodeWithoutLock(ctx, nodeID, wg)
```
If I remember correctly the main requirement is that connection and sending must be nonblocking. Correct?
If correct then we need a goroutine over connecting goroutines, otherwise, this call will be blocking.
If you meant by "nonblocking" something else then define it and I will reimplement it.

matejpavlovic commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,454 @@
+				}
+				conn.Stream = one.Stream
+				t.connsLock.Unlock()
+				t.logger.Log(logging.LevelDebug, "updated stream", "src", t.ownID, "nodeID", one.NodeID)
```
If we found the stream, let's try to close the old one (after releasing the lock).

dnkolegov replied:
Fixed

matejpavlovic commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,454 @@
+	info := conn.AddrInfo
+
+	t.host.Peerstore().AddAddrs(info.ID, info.Addrs, PermanentAddrTTL)
+
```
Release RLock here.

dnkolegov replied:
Fixed

matejpavlovic commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,454 @@
+	}
+}
+
+func (t *Transport) connectToNodeWithoutLock(ctx context.Context, nodeID types.NodeID) {
```
Pls also rename this (e.g. to `connectToNode`) - now it's using a lock again :-)

dnkolegov replied:
Fixed

matejpavlovic commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,450 @@
+
+	err = t.sendPayload(dest, out)
+	if err != nil {
+		t.connect(ctx, []types.NodeID{dest})
```
I'm a bit worried that, if the connection breaks and the new one takes some time to establish, there might be literally thousands of messages that fail to be sent in the meantime, each of them triggering a new connection attempt. All of these attempts (but one) will be closed almost immediately. Could that happen?

matejpavlovic replied:
In fact, nevermind, let's do this in a separate PR.

dnkolegov commented on pkg/net/libp2p/libp2p_transport.go:
```diff
@@ -0,0 +1,450 @@
+		delete(t.conns, nodeID)
+	}
+	t.connsLock.Unlock()
+}
```
Why do we need wg if we have the stopChan and it cannot exit until it gets a signal on this channel?
